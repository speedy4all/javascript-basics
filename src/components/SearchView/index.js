import React, { useCallback, useEffect, useRef } from "react";
import PropTypes from "prop-types";
import SearchButton from "./search-button";

export default function SearchView({ search, setSearch }) {
  const searchRef = useRef(null);

  useEffect(() => {
    searchRef.current.focus();
  }, []);

  useEffect(() => {
    console.log("render SearchView");
  }, [search, setSearch]);

  const onClick = useCallback(() => setSearch(""), [setSearch]);

  return (
    <div>
      <label htmlFor="searchInput">Search post</label>
      <input
        ref={searchRef}
        id="searchInput"
        onChange={({ target: { value } }) => setSearch(value)}
        value={search}
      />
      <SearchButton onClick={onClick} />
    </div>
  );
}

SearchView.propTypes = {
  search: PropTypes.string,
  setSearch: PropTypes.func,
};
