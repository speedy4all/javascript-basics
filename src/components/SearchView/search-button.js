import React, { memo } from "react";

function SearchButton({ onClick }) {
  return <button onClick={onClick} data-testid="clear-button">Clear</button>;
}

SearchButton.propTypes = {};

export default memo(SearchButton);
