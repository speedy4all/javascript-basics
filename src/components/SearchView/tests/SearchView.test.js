import { render, screen, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'
import SearchView from '../index';

const defaultProps = {
    setSearch: jest.fn(),
    search: 'new'
}

test('render correct searchView', () => {
    const { getByTestId } = render(<SearchView search={defaultProps.search} setSearch={defaultProps.setSearch}/>)
    const searchComponent = getByTestId('searchView-element')
    expect(searchComponent).toBeInTheDocument();
})

describe('Search input should', () => {
    it('not have any value initially', () => {
        render(<SearchView setSearch={defaultProps.setSearch}/>)
        const inputSearch = screen.getByTestId('searchView-input');
        expect(inputSearch.value).toBe('')
    })

    it('not have correct value from props', () => {
        render(<SearchView search="correct-search" setSearch={defaultProps.setSearch}/>)
        const inputSearch = screen.getByTestId('searchView-input');
        expect(inputSearch.value).toBe('correct-search')
    })

    it('not have correct value from props', () => {
        render(<SearchView search="correct-search" setSearch={defaultProps.setSearch}/>)
        const inputSearch = screen.getByTestId('searchView-input');
        expect(inputSearch.value).toBe('correct-search')
    })

    it('have correct input value after change', () => {
        render(<SearchView search="correct-search" setSearch={defaultProps.setSearch}/>)
        const inputSearch = screen.getByTestId('searchView-input');
        fireEvent.change(inputSearch, { target: { value: 'searchView'}})
        expect(defaultProps.setSearch).toHaveBeenCalledWith('searchView')
    })

    it('should render the correct snapshot', () => {
        const tree = renderer
            .create(<SearchView search={defaultProps.search} setSearch={defaultProps.setSearch}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    })

})
