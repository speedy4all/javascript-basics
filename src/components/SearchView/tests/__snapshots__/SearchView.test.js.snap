// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search input should should render the correct snapshot 1`] = `
<div
  data-testid="searchView-element"
>
  <label
    htmlFor="searchInput"
  >
    Search post
  </label>
  <input
    data-testid="searchView-input"
    id="searchInput"
    onChange={[Function]}
    value="new"
  />
  <button
    data-testid="clear-button"
    onClick={[Function]}
  >
    Clear
  </button>
</div>
`;
